src/noggit/scripting/scripting_tool.cpp

```diff
@@ -0,0 +1,711 @@
+    }
+
+    void scripting_tool::addStringList(char const* name, char const* value)
+    {
+
+      if (_string_arrays.find(name) == _string_arrays.end())
+      {
+        auto box = new QComboBox(this);
+        auto label = new QLabel(this);
+
+        connect(box, QOverload<int>::of(&QComboBox::activated), this, [=](auto index) {
+          set_json_unsafe<std::string>(name, box->itemText(index).toUtf8().constData());
+        });
+
+        box->addItem(value);
+        label->setText(name);
+
+        _string_arrays[name] = box;
+        _script_widgets.push_back(box);
+        _script_widgets.push_back(label);
+        _script_settings_layout->addRow(label, box);
+
+        // ensure there is at least one valid value in it
+        get_json_safe<std::string>(name, value);
+      }
+      else
+      {
+        auto box = _string_arrays[name];
+        box->addItem(value);
+
+        // we found the last selection, so change the index for that.
+        if (get_json_safe<std::string>(name, "") == value)
+        {
+          box->setCurrentIndex(box->count() - 1);
+        }
+      }
+    }
+
+    void scripting_tool::addString(char const* name, char const* def)
+    {
+      auto defstr = def == nullptr ? "" : def;
+      auto tline = new QLineEdit(this);
+      auto label = new QLabel(this);
+      label->setText(name);
+      connect(tline, &QLineEdit::textChanged, this, [=](auto text) {
+        set_json_unsafe<std::string>(name, text.toUtf8().constData());
+      });
+      _script_widgets.push_back(label);
+      _script_widgets.push_back(tline);
+      _script_settings_layout->addRow(label, tline);
+      tline->setText(get_json_safe<std::string>(name, defstr).c_str());
+    }
+
+    void scripting_tool::removeScriptWidgets()
+    {
+      for (auto& widget : _script_widgets)
+      {
+        _script_settings_layout->removeWidget(widget);
+        delete widget;
+      }
+
+      _script_widgets.clear();
+      _string_arrays.clear();
+    }
+
+    void scripting_tool::select_profile(int profile)
+    {
+      removeScriptWidgets();
+      clearDescription();
+      cur_tool = this;
+      cur_profile = _profile_selection->itemText(profile).toUtf8().constData();
+
+      auto n = selected_script_name();
+      if (!_json.contains(n))
+      {
+        _json[n] = json();
+      }
+
+      _json[n][CUR_PROFILE_PATH] = cur_profile;
+
+      initialize_radius();
+
+      select_script(get_selected_script());
+      cur_tool = nullptr;
+    }
+
+    void scripting_tool::sendUpdate(
+        World* world,
+        math::vector_3d pos_in,
+        noggit::camera* cam,
+        float dt,
+        bool left_mouse,
+        bool right_mouse,
+        bool holding_shift,
+        bool holding_ctrl,
+        bool holding_alt,
+        bool holding_space)
+    {
+      cur_tool = this;
+      script_context ctx(world, pos_in, brushRadius(), innerRadius(), cam, holding_alt, holding_shift, holding_ctrl, holding_space, dt);
+      set_ctx(&ctx);
```

Comment:
script_context seems to be only allowed to be != nullptr within the scope of this function.
consider using a RAII pattern similar to the other `scoped_*` contexts noggit already has (opengl)